File: src/patterns/structural/composite/coumpoundgraphic.nim
# internal
import
    patterns/structural/composite/[
        circle,
        graphic,
        dot
    ]


type
    CoumpoundGraphic* = ref object
        children: seq[GraphicImpl]


proc newCoumpoundGraphic*(): CoumpoundGraphic =
    return CoumpoundGraphic()

proc add*(self: CoumpoundGraphic, child: Graphic) =
    self.children.add(child)

proc move*(self: CoumpoundGraphic, x,y: float): bool =
    for child in self.children:
        discard child.move(x,y)

proc draw*(self: CoumpoundGraphic): bool =
    for child in self.children:
        discard child.draw()

File: src/patterns/structural/composite/coumpoundgraphic.go
package composite


type CoumpoundGraphic struct {
	children []Graphic
}


func NewCoumpoundGraphic() *CoumpoundGraphic {
	return &CoumpoundGraphic{}
}

func (self *CoumpoundGraphic) AddChild(child Graphic) {
	self.children = append(self.children, child)
}

func (self *CoumpoundGraphic) Move(x,y float32) bool {
	for _, child := range self.children {
		child.Move(x,y)
	}

	return true
}

func (self *CoumpoundGraphic) Draw() bool {
	for _, child := range self.children {
		child.Draw()
	}

	return true
}